## 1. Introduction
<p><img src="https://assets.datacamp.com/production/project_1197/img/google_play_store.png" alt="Google Play logo"></p>
<p>Mobile apps are everywhere. They are easy to create and can be very lucrative from the business standpoint. Specifically, Android is expanding as an operating system and has captured more than 74% of the total market<sup><a href="https://www.statista.com/statistics/272698/global-market-share-held-by-mobile-operating-systems-since-2009">[1]</a></sup>. </p>
<p>The Google Play Store apps data has enormous potential to facilitate data-driven decisions and insights for businesses. In this notebook, we will analyze the Android app market by comparing ~10k apps in Google Play across different categories. We will also use the user reviews to draw a qualitative comparision between the apps.</p>
<p>The dataset you will use here was scraped from Google Play Store in September 2018 and was published on <a href="https://www.kaggle.com/lava18/google-play-store-apps">Kaggle</a>. Here are the details: <br>
<br></p>
<div style="background-color: #efebe4; color: #05192d; text-align:left; vertical-align: middle; padding: 15px 25px 15px 25px; line-height: 1.6;">
    <div style="font-size:20px"><b>datasets/apps.csv</b></div>
This file contains all the details of the apps on Google Play. There are 9 features that describe a given app.
<ul>
    <li><b>App:</b> Name of the app</li>
    <li><b>Category:</b> Category of the app. Some examples are: ART_AND_DESIGN, FINANCE, COMICS, BEAUTY etc.</li>
    <li><b>Rating:</b> The current average rating (out of 5) of the app on Google Play</li>
    <li><b>Reviews:</b> Number of user reviews given on the app</li>
    <li><b>Size:</b> Size of the app in MB (megabytes)</li>
    <li><b>Installs:</b> Number of times the app was downloaded from Google Play</li>
    <li><b>Type:</b> Whether the app is paid or free</li>
    <li><b>Price:</b> Price of the app in US$</li>
    <li><b>Last Updated:</b> Date on which the app was last updated on Google Play </li>

</ul>
</div>
<div style="background-color: #efebe4; color: #05192d; text-align:left; vertical-align: middle; padding: 15px 25px 15px 25px; line-height: 1.6;">
    <div style="font-size:20px"><b>datasets/user_reviews.csv</b></div>
This file contains a random sample of 100 <i>[most helpful first](https://www.androidpolice.com/2019/01/21/google-play-stores-redesigned-ratings-and-reviews-section-lets-you-easily-filter-by-star-rating/)</i> user reviews for each app. The text in each review has been pre-processed and passed through a sentiment analyzer.
<ul>
    <li><b>App:</b> Name of the app on which the user review was provided. Matches the `App` column of the `apps.csv` file</li>
    <li><b>Review:</b> The pre-processed user review text</li>
    <li><b>Sentiment Category:</b> Sentiment category of the user review - Positive, Negative or Neutral</li>
    <li><b>Sentiment Score:</b> Sentiment score of the user review. It lies between [-1,1]. A higher score denotes a more positive sentiment.</li>

</ul>
</div>
<p>From here on, it will be your task to explore and manipulate the data until you are able to answer the three questions described in the instructions panel.<br></p>

**CODE BEGINS HERE**

Import the relevant modules and load the two csv files as DataFrame.

In [2]:
import pandas as pd

In [3]:
apps = pd.read_csv('datasets/apps.csv')
reviews = pd.read_csv('datasets/user_reviews.csv')

Let's get a thorough insight on the dataset using the .info() and .head() methods on the two DataFrames.

In [4]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9659 entries, 0 to 9658
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   App           9659 non-null   object 
 1   Category      9659 non-null   object 
 2   Rating        8196 non-null   float64
 3   Reviews       9659 non-null   int64  
 4   Size          8432 non-null   float64
 5   Installs      9659 non-null   object 
 6   Type          9659 non-null   object 
 7   Price         9659 non-null   float64
 8   Last Updated  9659 non-null   object 
dtypes: float64(3), int64(1), object(5)
memory usage: 679.3+ KB


In [5]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   App                 64295 non-null  object 
 1   Review              37427 non-null  object 
 2   Sentiment Category  37432 non-null  object 
 3   Sentiment Score     37432 non-null  float64
dtypes: float64(1), object(3)
memory usage: 2.0+ MB


In [6]:
reviews.head()

App                                             Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment Category  Sentiment Score  
0           Positive             1.00  
1           Positive             0.25  
2                NaN              NaN  
3           Positive             0.40  
4           Positive             1.00

The .sample() method is used to generate a random sample of data from the DataFrame. It is better than the .head() method as more diverse data can be examined.

In [7]:
apps.sample(10)

App         Category  \
3341                        My Android Device S/W & H/W           FAMILY   
8719                         Flag Of European Union LWP  PERSONALIZATION   
7525               Cursive Writing Wizard - Handwriting           FAMILY   
6042                                      CA Mobile OTP            TOOLS   
8170                                           EF Forms         BUSINESS   
498   Chatting - Free chat, random chat, boyfriend, ...           DATING   
7398                            Salah Widget (DK+Malmo)        LIFESTYLE   
5023                                          Upohar BD         BUSINESS   
7454                                         DM airdisk            TOOLS   
2571                      Fraction Calculator Plus Free            TOOLS   

      Rating  Reviews  Size    Installs  Type  Price       Last Updated  
3341     4.5      681   2.8     50,000+  Free    0.0  November 20, 2015  
8719     4.3       88   4.2      1,000+  Free    0.0   January 24, 2018  
7525     4.0     3745  31.0  1,000,000+  Free    0.0   December 3, 2017  
6042     3.3      688   5.5    100,000+  Free    0.0  September 5, 2017  
8170     5.0        2  23.0         50+  Free    0.0      July 24, 2018  
498      4.2     2506   6.1    500,000+  Free    0.0      June 15, 2017  
7398     4.5      532   6.4     10,000+  Free    0.0      June 28, 2015  
5023     3.4        9   1.7      1,000+  Free    0.0    October 8, 2016  
7454     2.8       11  11.0      1,000+  Free    0.0     August 5, 2016  
2571     4.5   148506  11.0  5,000,000+  Free    0.0      July 15, 2018

The first task was to convert the values on 'Installs' column in the dataset to an integer value. Originally, the 'Installs' column contained String values.


To convert the column to integer, we first need to remove special signs, like the '+' and ',' signs present in values on the 'Installs' column. We did that using the replace() method on the Series. The result was a string, but without the operator signs. We then used the to_numeric() function from Pandas to convert the Series from string (referred to in pandas as Object data type) to int64 type.

In [8]:
# Removing special characters from the 'Installs' column.

chars_to_replace = [',', '+']

for char in chars_to_replace:
    apps['Installs'] = apps['Installs'].str.replace(char, '', regex=False)
apps['Installs'] = pd.to_numeric(apps['Installs'], errors='raise')

We can see that the Installs column is now of data type int64.

In [9]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9659 entries, 0 to 9658
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   App           9659 non-null   object 
 1   Category      9659 non-null   object 
 2   Rating        8196 non-null   float64
 3   Reviews       9659 non-null   int64  
 4   Size          8432 non-null   float64
 5   Installs      9659 non-null   int64  
 6   Type          9659 non-null   object 
 7   Price         9659 non-null   float64
 8   Last Updated  9659 non-null   object 
dtypes: float64(3), int64(2), object(4)
memory usage: 679.3+ KB


Again, let's have a look at the modified apps DataFrame.

In [10]:
apps.sample(10)

App             Category  Rating  Reviews  \
5628           Theme for Android 7.0      PERSONALIZATION     NaN        0   
6697                         iReadMe         PRODUCTIVITY     5.0        8   
646   GRE Prep & Practice by Magoosh            EDUCATION     4.4     3963   
9485          FN FAL rifle explained  BOOKS_AND_REFERENCE     NaN        1   
6267             Best CG Photography               FAMILY     NaN        1   
6310                 DRAGON QUEST IV               FAMILY     4.7     1647   
9361                      FK Liepaja               SPORTS     NaN        4   
9400  Scratch-Off Guide for FL Lotto               FAMILY     4.6       25   
3759  Grand Theft Auto V: The Manual               FAMILY     4.1   182103   
5040                 BANGLA TV 3G/4G               SPORTS     4.1      142   

      Size  Installs  Type  Price       Last Updated  
5628   NaN      5000  Free   0.00   October 18, 2017  
6697  22.0       100  Free   0.00      March 6, 2018  
646    NaN    100000  Free   0.00       June 6, 2018  
9485   7.3        10  Paid   6.49  September 6, 2015  
6267   2.5       500  Free   0.00      June 24, 2015  
6310  16.0     10000  Paid  14.99  February 28, 2017  
9361  26.0       100  Free   0.00     March 17, 2018  
9400   5.6      5000  Free   0.00       July 4, 2018  
3759   6.9   5000000  Free   0.00      July 31, 2018  
5040   5.3    100000  Free   0.00       June 7, 2018

We can also see that the Date is of type 'str'. There is a special data type in Python, called datetime, which is used to process date/ time values. So, let's look at converting the values from the 'Last Updated' column to type datetime.

pandas has a valuable function called to_datetime(), which is similar to the to_numeric() function we used earlier. However, in this function, we need to specify the format of the date/ time in the present column. For this, we can use metacharacters. Metacharacters are special characters that represent certain values in Python.

I looked up the key for metacharacters at the strftime docs available at https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior.

In [11]:
date = pd.to_datetime(apps['Last Updated'], format = '%B %d, %Y')
apps['Last Updated'] = date

In [12]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9659 entries, 0 to 9658
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   App           9659 non-null   object        
 1   Category      9659 non-null   object        
 2   Rating        8196 non-null   float64       
 3   Reviews       9659 non-null   int64         
 4   Size          8432 non-null   float64       
 5   Installs      9659 non-null   int64         
 6   Type          9659 non-null   object        
 7   Price         9659 non-null   float64       
 8   Last Updated  9659 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(3), int64(2), object(3)
memory usage: 679.3+ KB


For the next specified task, we were to find the number of apps in each category, the average price, and the average rating. The answer was to be saved as a DataFrame called app_category_info, with the four columns named: Category, Number of apps, Average price, Average rating.

Now, we need to group the rows by 'Category' as specified in the question. Then, we count the number of apps using the count() method on the GroupBy object generated using the groupby() method. Similarly, we use the agg() method to calculate the averages for the price and rating for all groups in the category. 

In [16]:
no_apps_df = apps.groupby('Category')['App'].count()
avg_price_df = apps.groupby('Category')['Price'].agg('mean')
avg_rating_df = apps.groupby('Category')['Rating'].agg('mean')

In [17]:
app_category_info = pd.DataFrame(data = {
                     'Number of apps':no_apps_df, 
                     'Average price':avg_price_df, 
                     'Average rating':avg_rating_df})

In [18]:
# DATAFRAME CREATED, #2 DONE

app_category_info.reset_index()

Category  Number of apps  Average price  Average rating
0        ART_AND_DESIGN              64       0.093281        4.357377
1     AUTO_AND_VEHICLES              85       0.158471        4.190411
2                BEAUTY              53       0.000000        4.278571
3   BOOKS_AND_REFERENCE             222       0.539505        4.344970
4              BUSINESS             420       0.417357        4.098479
5                COMICS              56       0.000000        4.181481
6         COMMUNICATION             315       0.263937        4.121484
7                DATING             171       0.160468        3.970149
8             EDUCATION             119       0.150924        4.364407
9         ENTERTAINMENT             102       0.078235        4.135294
10               EVENTS              64       1.718594        4.435556
11               FAMILY            1832       1.309967        4.179664
12              FINANCE             345       8.408203        4.115563
13       FOOD_AND_DRINK             112       0.075714        4.172340
14                 GAME             959       0.296465        4.247368
15   HEALTH_AND_FITNESS             288       0.223438        4.243033
16       HOUSE_AND_HOME              74       0.000000        4.150000
17   LIBRARIES_AND_DEMO              84       0.011786        4.178125
18            LIFESTYLE             369       6.398022        4.093355
19  MAPS_AND_NAVIGATION             131       0.205725        4.036441
20              MEDICAL             395       2.520759        4.166552
21   NEWS_AND_MAGAZINES             254       0.015669        4.121569
22            PARENTING              60       0.159667        4.300000
23      PERSONALIZATION             376       0.400213        4.332215
24          PHOTOGRAPHY             281       0.420925        4.157414
25         PRODUCTIVITY             374       0.670936        4.183389
26             SHOPPING             202       0.027129        4.230000
27               SOCIAL             239       0.066820        4.247291
28               SPORTS             325       0.307692        4.216154
29                TOOLS             827       0.323156        4.039554
30     TRAVEL_AND_LOCAL             219       0.228082        4.069519
31        VIDEO_PLAYERS             163       0.064172        4.044595
32              WEATHER              79       0.410380        4.243056

For the third task, we were to find the top 10 free FINANCE apps having the highest average sentiment score. A DataFrame named top_10_user_feedback with the columns App and Sentiment Score was to be created, with the average sentiment scores sorted in descending order.

We can see from the above output that there are a total of 345 apps in the 'Finance' category.

Now, to find the average sentiment scores for each app, we need to first have a dataframe consisting of all the apps, their categories and the Sentiment Scores. This means we need to access values from both apps and reviews DataFrames. This can be done by joining the two dataframes on the 'App' column.

In [20]:
score_apps_table = apps.merge(reviews, on='App', how='inner')

In [21]:
score_apps_table.sample(5)

App            Category  Rating  Reviews  \
10800                  GRE Flashcards           EDUCATION     4.5    13791   
45507  Golf GPS Rangefinder: Golf Pad              SPORTS     4.6    13098   
45677  Fantasy Football Manager (FPL)              SPORTS     4.4    63650   
18051    Couch to 10K Running Trainer  HEALTH_AND_FITNESS     4.6    10445   
44842                      Championat              SPORTS     4.3    36490   

       Size  Installs  Type  Price Last Updated  \
10800   NaN    500000  Free    0.0   2018-06-06   
45507   NaN   1000000  Free    0.0   2018-07-28   
45677   4.6   1000000  Free    0.0   2018-08-03   
18051  48.0    500000  Free    0.0   2018-07-02   
44842   NaN   1000000  Free    0.0   2018-07-25   

                                                  Review Sentiment Category  \
10800  Amazing app, gives tips words meaning relevant...           Positive   
45507  I bought tags easy use. It take time get tags ...           Positive   
45677  Entertaining,, though ads disruptive, disturbi...           Negative   
18051  Does good job interval training I wish I could...           Positive   
44842                     Tournament Table does not work            Neutral   

       Sentiment Score  
10800         0.616667  
45507         0.567917  
45677        -0.280000  
18051         0.333333  
44842         0.000000

Let's now create a DataFrame that contains only the free apps from the FINANCE category, as required by the question.

In [22]:
finance_scores_table = score_apps_table[(score_apps_table['Category'] == 'FINANCE') & (score_apps_table['Type'] == 'Free')]

The DataFrame contains 2200 columns, even though we found out above that there are only 328 apps in the 'Finance' category. This is because the same app contains multiple reviews.

In [43]:
top_10_user_feedback = pd.DataFrame(finance_scores_table.groupby('App')['Sentiment Score'].agg('mean'))

In [44]:
top_10_user_feedback.groupby('App')['Sentiment Score'].agg('mean').sort_values(ascending = False)

App
BBVA Spain                                         0.515086
Associated Credit Union Mobile                     0.388093
BankMobile Vibe App                                0.353455
A+ Mobile                                          0.329592
Current debit card and app made for teens          0.327258
BZWBK24 mobile                                     0.326883
Even - organize your money, get paid early         0.283929
Credit Karma                                       0.270052
Fortune City - A Finance App                       0.266966
Branch                                             0.264230
ACE Elite                                          0.252171
Digit Save Money Automatically                     0.250246
CNBC: Breaking Business News & Live Market Data    0.247704
CreditWise from Capital One                        0.239454
Expense IQ Money Manager                           0.227506
BBVA Compass Banking                               0.205590
Bank of America Mobile Banking      

In [46]:
top_10_user_feedback = pd.DataFrame(top_10_user_feedback.groupby('App')['Sentiment Score'].agg('mean').sort_values(ascending=False)[:10])

In [47]:
top_10_user_feedback

Sentiment Score
App                                                        
BBVA Spain                                         0.515086
Associated Credit Union Mobile                     0.388093
BankMobile Vibe App                                0.353455
A+ Mobile                                          0.329592
Current debit card and app made for teens          0.327258
BZWBK24 mobile                                     0.326883
Even - organize your money, get paid early         0.283929
Credit Karma                                       0.270052
Fortune City - A Finance App                       0.266966
Branch                                             0.264230